### ISSUE TITLE
🐛 [cloud_firestore] Internal error in Cloud Firestore causing the flutter app to crash 

### ISSUE THREAD

**Issue by sumittiware** [firebase/flutterfire#11674]
## Bug report

**Describe the bug**
Internal error in Cloud Firestore causing the app to crash, please refer the following stacktrace.

Got this issue in Crashlytics

```
Fatal Exception: java.lang.RuntimeException: Internal error in Cloud Firestore (24.4.3).
       at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$3(AsyncQueue.java:545)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8680)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:567)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)

Caused by java.lang.UnsupportedOperationException:
       at java.util.AbstractMap.put(AbstractMap.java:209)
       at java.util.AbstractMap.putAll(AbstractMap.java:281)
       at com.google.firebase.firestore.local.LocalDocumentsView.populateOverlays(LocalDocumentsView.java:361)
       at com.google.firebase.firestore.local.LocalDocumentsView.getNextDocuments(LocalDocumentsView.java:344)
       at com.google.firebase.firestore.local.IndexBackfiller.writeEntriesForCollectionGroup(IndexBackfiller.java:136)
       at com.google.firebase.firestore.local.IndexBackfiller.writeIndexEntries(IndexBackfiller.java:119)
       at com.google.firebase.firestore.local.IndexBackfiller.lambda$backfill$0(IndexBackfiller.java:105)
       at com.google.firebase.firestore.local.SQLitePersistence.runTransaction(SQLitePersistence.java:228)
       at com.google.firebase.firestore.local.IndexBackfiller.backfill(IndexBackfiller.java:105)
       at com.google.firebase.firestore.local.IndexBackfiller$Scheduler.lambda$scheduleBackfill$0(IndexBackfiller.java:92)
       at com.google.firebase.firestore.util.AsyncQueue$DelayedTask.handleDelayElapsed(AsyncQueue.java:145)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:487)
       at java.util.concurrent.FutureTask.run(FutureTask.java:264)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:235)
       at java.lang.Thread.run(Thread.java:1012)
```

**Comment by danagbemava-nc** [comment]
Hi @sumittiware, at what point in your app do you experience this crash?

How many instances of this crash do you have? 

What version of cloud_firestore are you using?

**Comment by sumittiware** [comment]
Hi @danagbemava-nc, thank you so much for the reply

1. As per the crashlytics logs, all the crashes are registered on a particular screen where here we are listening to the firestore stream and performing the CRUD operations on data in the same collection.
2. As per crashlyics all in last 90 days, 173 crashes are registered. All the instances are having same logs on the crashlytics dashboard.
3. cloud_firestore: ^4.1.0.

Since no dart stack-track is attached in the logs, we couldn't understand what is causing this crash, need help in debugging this issue.

**Comment by danagbemava-nc** [comment]
Hi @sumittiware, can you share the operations that you perform on that page? 

Are you writing any complex queries?

**Comment by sumittiware** [comment]
We are mostly using the read operations here, since the error is with the put method it might be because of the write query only, we have 3 such queries 2 are very simple we are just setting the jsin object and one is a but complex attaching the image here.
<img width="533" alt="Screenshot 2023-10-12 at 9 17 15 AM" src="https://github.com/firebase/flutterfire/assets/64718675/6f9cfd08-1b5a-403d-a136-f4a936c29e4e">

**Comment by darshankawar** [comment]
@sumittiware 
Can you check if [this](https://stackoverflow.com/a/73415635) solution helps in your case or not ?

**Comment by sumittiware** [comment]
@darshankawar Thank you for the response, I will add this changes and monitor.

**Comment by sumittiware** [comment]
Hi @darshankawar, I have made the above updates, but I still got a crash in the Firebase Crashlytics. I couldn’t reproduce in local & there is only single instance of it.

**Comment by sumittiware** [comment]
I think I might have identified the issue. This has something to do with the android implementation of the firestore dependency.

From the above stack-trace we can identify that the the code is failing because it is trying to call put method on the abstract map, in java this method is not implemented.

at java.util.AbstractMap.put(AbstractMap.java:209)

Now, as we navigated down the stack-trace we have found that, there is a condition in the firestore-android-SDK that there a condition where the overlays is initialised with the Abstract map and passed down the code.

```
 Map<DocumentKey, Overlay> overlays =
        count - docs.size() > 0
            ? documentOverlayCache.getOverlays(
                collectionGroup, offset.getLargestBatchId(), count - docs.size())
            : Collections.emptyMap();
```
in above code emptyMap() returns an EmptyMap() which does not have put method implemented which is calling put for AbstractMap and in-turn throwing the UnsupportedOperationException.
 
<img width="718" alt="image" src="https://github.com/firebase/flutterfire/assets/64718675/59384b8c-1b6e-4049-b1f8-9ceeff542e86">

https://github.com/firebase/firebase-android-sdk/blob/18fa5f309a9eb86198fa6e59b96987a66b99ef89/firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java#L323
